### PR TITLE
[TS-244] Added activities route

### PR DESCRIPTION
**Pull request by tungleu**
added endpoint to get recent activities 
`GET /actitivities/:userId`
image

### CODE REVIEW

File: src/controllers/activities.controller.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+class ActivitiesController {
+	private userInfoService: UserInfoService;
+	constructor() {
+		this.userInfoService = new UserInfoService(UserInfoCollection);
+	}
+
+	async getActivities(req: Request, res: Response) {
+		const userId = new mongoose.Types.ObjectId(req.params.userId);
+		const userInfo = await this.userInfoService.findUserInfo({ userId: userId });
+		let account = {};
```
Comment (bowe99) [inline src/controllers/activities.controller.ts:17]: This should be renamed to activities

File: src/controllers/activities.controller.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+				headers: { Authorization: `Bearer ` + alpacaToken },
+			});
+			account = response.data;
+			return res.send(messages.successMessage("success", "activities", account))
+		}
+		catch (error: any) {
+			 res.send(messages.internalError(error.response.data.message));
+		}
+		}
+		res.send(messages.internalError("User hasn't linked any Alpaca account"));
```
Comment (bowe99) [inline src/controllers/activities.controller.ts:31]: This code is failing Sonar because of duplication maybe you can clean it up a little

**Review by bowe99 — CHANGES_REQUESTED** [review]

File: src/routes/v1/activities.route.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import express, { Express } from "express";
+import ActivitiesController from "../../controllers/activities.controller";
+import requireUser from "../../middleware/requireUser";
```
Comment (bowe99) [inline src/routes/v1/activities.route.ts:3]: This is not being used you can remove it

**Review by bowe99 — CHANGES_REQUESTED** [review]

File: src/db/service/alpaca.service.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+		infoType: string,
+		alpacaToken: string
+	) {
+		try {
+			let alpacaInfo = {};
+			const response = await axios.get(endPoint, {
+				headers: { Authorization: `Bearer ` + alpacaToken },
+			});
+			alpacaInfo = response.data;
+			return res.send(messages.successMessage("success", infoType, alpacaInfo));
```
Comment (kennguyen0303) [inline src/db/service/alpaca.service.ts:20]:
```suggestion
			const {data}= await axios.get(endPoint, {
				headers: { Authorization: `Bearer ` + alpacaToken },
			});
			return res.send(messages.successMessage("success", infoType, data));
```

Save you one line and one variable

File: src/controllers/positions.controller.ts [hunk 5]
```diff
@@ -2,33 +2,25 @@ import { Request, Response } from "express";
...
-		let account = {};
 		if (userInfo?.alpacaToken) {
-			try{
-			const alpacaToken = userInfo.alpacaToken;
-			const response = await axios.get("/positions", {
-				headers: { Authorization: `Bearer ` + alpacaToken },
-			});
-			account = response.data;
-			return res.send(messages.successMessage("success", "positions", account))
+			return this.alpacaService.getInfo(req, res, "/positions", "positions", userInfo.alpacaToken);
```
Comment (kennguyen0303) [inline src/controllers/positions.controller.ts:20]:
```suggestion
			return await this.alpacaService.getInfo(req, res, "/positions", "positions", userInfo.alpacaToken);
```

missing `await`

File: src/controllers/positions.controller.ts [hunk 6]
```diff
@@ -2,33 +2,25 @@ import { Request, Response } from "express";
...
+		} else {
+			res.send(messages.internalError("User hasn't linked any Alpaca account"));
```
Comment (kennguyen0303) [inline src/controllers/positions.controller.ts:22]:
```suggestion
			return res.send(messages.internalError("User hasn't linked any Alpaca account"));
```

Missing `return` for convention

File: src/controllers/activities.controller.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+	constructor() {
+		this.userInfoService = new UserInfoService(UserInfoCollection);
+		this.alpacaService = new AlpacaService();
+	}
+
+	async getActivities(req: Request, res: Response) {
+		const userId = new mongoose.Types.ObjectId(req.params.userId);
+		const userInfo = await this.userInfoService.findUserInfo({ userId: userId });
+		if (userInfo?.alpacaToken) {
+			return this.alpacaService.getInfo(
```
Comment (kennguyen0303) [inline src/controllers/activities.controller.ts:19]:
```suggestion
			return await this.alpacaService.getInfo(
```

missing `await`

File: src/controllers/account.controller.ts [hunk 8]
```diff
@@ -2,33 +2,25 @@ import { Request, Response } from "express";
...
-			try{
-			const alpacaToken = userInfo.alpacaToken;
-			const response = await axios.get("/account", {
-				headers: { Authorization: `Bearer ` + alpacaToken },
-			});
-			account = response.data;
-			return res.send(messages.successMessage("success", "account", account))
+			return this.alpacaService.getInfo(req, res, "/account", "account", userInfo.alpacaToken);
+		} else {
+			res.send(messages.internalError("User hasn't linked any Alpaca account"));
```
Comment (kennguyen0303) [inline src/controllers/account.controller.ts:22]:
```suggestion
			return res.send(messages.internalError("User hasn't linked any Alpaca account"));
```

missing `return`

File: src/controllers/account.controller.ts [hunk 9]
```diff
@@ -2,33 +2,25 @@ import { Request, Response } from "express";
...
-		let account = {};
 		if (userInfo?.alpacaToken) {
-			try{
-			const alpacaToken = userInfo.alpacaToken;
-			const response = await axios.get("/account", {
-				headers: { Authorization: `Bearer ` + alpacaToken },
-			});
-			account = response.data;
-			return res.send(messages.successMessage("success", "account", account))
+			return this.alpacaService.getInfo(req, res, "/account", "account", userInfo.alpacaToken);
```
Comment (kennguyen0303) [inline src/controllers/account.controller.ts:20]:
```suggestion
			return await this.alpacaService.getInfo(req, res, "/account", "account", userInfo.alpacaToken);
```

missing `await`

**Review by kennguyen0303 — CHANGES_REQUESTED** [review]
Great work, only 2 small things you might have known. I think having `await` and `return` is necessary. Although there might be no issue on your method, it triggers unexpected behavior sometimes if you work on a different feature. 

Food for thoughts: `await` will tell NodeJS to mark that line, run that function in synchronous behavior and wait for a return; As a result, if your workflow contains only one method, it won't make any difference. Without `await`, nodejs will jump over your call without waiting for the return/response, possibly go to the next lines of code (if there any).

Thanks @bowe99 for noticing the unused imports, I would have missed them for sure.

**Review by kennguyen0303 — COMMENTED** [review]
I would like to have a unit test for the service before this is merged. Will it be possible?

**Review by kennguyen0303 — DISMISSED** [review]
lgtm!

**Review by kennguyen0303 — APPROVED** [review]

**Comment by kennguyen0303** [comment]
@bowe99 need your review here since you requested some changes.

**Review by kennguyen0303 — APPROVED** [review]

**Review by GadoliniumPotassium — APPROVED** [review]

**Pull request merged**